Fix connection reuse + port pre-flight

Mzack9999 opened the pull request:
## Proposed changes

Speeds up scans with a TCP port pre-flight, per-host rate limiting, per-host HTTP client pooling and sharding, connection reuse tracking, and automatic HTTP to HTTPS port correction, all behind CLI flags.

Trimmed the `tests` workflow wall-clock ~60m -> ~35m by running `-race` on ubuntu only, folding `go vet` into the lint gate, and adding build caching to the integration/functional jobs, with the same coverage and fewer runner-minutes.

## Checklist
- [ ] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Preflight TCP port scan to detect and filter responsive targets before execution
  * Per-host rate limiting, per-host HTTP client pooling, and HTTP client sharding
  * Connection reuse tracking with per-host statistics and a connection-reuse policy
  * Automatic HTTP→HTTPS port-mismatch detection and correction
  * New CLI flags to toggle these optimizations

* **Tests**
  * Updated leak-suppression rules to ignore background LRU TTL goroutines

* **Chores**
  * Explicit dependency declared in module file
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Mzack9999 committed:
.

Mzack9999 committed:
.

Mzack9999 committed:
adding pre-flight

Mzack9999 committed:
add conn reuse

Mzack9999 committed:
Merge branch 'dev' into feat-6702-http-client

Mzack9999 committed:
making pre-flight optional

Mzack9999 committed:
adding per-host httpclient + ratelimit

Mzack9999 committed:
add conn reuse stats

Mzack9999 committed:
ignoring leak in lru

Mzack9999 committed:
adding global connection pooling with sharding

Mzack9999 committed:
lint

Ice3man543 commented:
@coderabbitai review

Ice3man543 committed:
Merge branch 'dev' of https://github.com/projectdiscovery/nuclei into feat-6702-http-client

dwisiswant0 commented on pkg/protocols/http/httpclientpool/perhost_ratelimit_pool.go:
```diff
@@ -0,0 +1,511 @@
+
+func (p *PerHostRateLimitPool) GetOrCreate(
+	host string,
+) (*ratelimit.Limiter, error) {
```
Heavy read op here. (`GetForTarget` -> `GetOrCreate`)

Maybe we should switch from LRU to adaptive W-TinyLFU (w/ [`github.com/maypok86/otter/v2`](https://pkg.go.dev/github.com/maypok86/otter/v2) - already introduced in #6630) to get lock-free access?

Thus, no more leaks & growing `goleak.Ingore*` list.

dwisiswant0 commented on pkg/protocols/http/httpclientpool/perhost_ratelimit_pool.go:
```diff
@@ -0,0 +1,511 @@
+	if len(entry.requestTimestamps) > 100 {
+		// Keep only last 100 timestamps
+		entry.requestTimestamps = entry.requestTimestamps[len(entry.requestTimestamps)-100:]
+	}
```
Implicit re-allocs (sliding window) here.

Since this is hardcapped at 100, maybe a circular buffer would be a better fit than a ring buffer for `requestTimestamps`?

dwisiswant0 reviewed (changes requested):
There's only one concern for correctness, else is just nits (perf, bottlenecks view).

Mzack9999 committed:
Merge remote-tracking branch 'origin/dev' into feat-6702-http-client

Mzack9999 committed:
addressing comments

Mzack9999 committed:
fix lint: disambiguate retryablehttp URL access

retryablehttp.Request embeds *urlutil.URL, so .URL.String() resolved to
the wrong type and tripped staticcheck QF1008. Use .Request.URL to read
the underlying http.Request URL for rate limiting and pooling keys.

Mzack9999 committed:
fix race

Mzack9999 committed:
regression harness

Mzack9999 committed:
ci tuning

Ice3man543 reviewed (approved):
lgtm! tested it out and gave quite a lot of improvements over baseline. looks ready to merge i think. 👍

ehsandeep reviewed (changes requested):
Merge conflict after - https://github.com/projectdiscovery/nuclei/pull/7301

Mzack9999 committed:
address review comments

- preflight: resolve via MetaInput.Target()/CustomIP, gate banner on Silent
- http: drop dead Priority 4 branch in AnalyzeConnectionReuse

Mzack9999 committed:
reconcile dev

Mzack9999 committed:
add tests

Mzack9999 committed:
host stats

Mzack9999 committed:
reuse timing
